src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -199,27 +188,137 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig, protected
     val fileInserter = storage.objects().insert(bucketName, storageObject, content)
     // Enable direct media upload
     fileInserter.getMediaHttpUploader().setDirectUploadEnabled(true)
-    Future {
-      executeGoogleRequest(fileInserter) //returns a StorageObject
-    }.recoverWith {
-      case e: GoogleJsonResponseException => throw CallToGoogleApiFailedException(googleProject, bucketName, e.getStatusCode, e.getDetails.getMessage)
-    } map { storageObject =>
+    executeGoogleRequestAsync(googleProject, bucketName, fileInserter) map { storageObject =>
       StorageObjectResponse(storageObject.getName, storageObject.getBucket, storageObject.getTimeCreated.toString)
     }
   }
 
   /* Delete a cluster within the google project */
   def deleteCluster(googleProject: String, clusterName: String)(implicit executionContext: ExecutionContext): Future[Unit] = {
-    Future {
-      val request = dataproc.projects().regions().clusters().delete(googleProject, dataprocConfig.dataprocDefaultZone, clusterName)
-      try {
-        executeGoogleRequest(request)
-      } catch {
-        case e:GoogleJsonResponseException =>
-          if(e.getStatusCode!=404)
-            throw CallToGoogleApiFailedException(googleProject, clusterName, e.getStatusCode, e.getDetails.getMessage)
+    val request = dataproc.projects().regions().clusters().delete(googleProject, dataprocConfig.dataprocDefaultRegion, clusterName)
+    executeGoogleRequestAsync(googleProject, clusterName, request).recover {
+      // treat a 404 error as a successful deletion
+      case CallToGoogleApiFailedException(_, _, 404, _) => ()
+    }.void
+  }
+
+  override def getClusterStatus(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[LeoClusterStatus] = {
+    getCluster(googleProject, clusterName).map { cluster =>
+      LeoClusterStatus.withNameIgnoreCase(cluster.getStatus.getState)
+    }
+  }
+
+  override def getClusterMasterInstanceIp(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[Option[String]] = {
+    // OptionT is handy when you potentially want to deal with Future[A], Option[A],
+    // Future[Option[A]], and A all in the same flatMap!
+    //
+    // Legend:
+    // - OptionT.pure turns an A into an OptionT[F, A]
+    // - OptionT.liftF turns an F[A] into an OptionT[F, A]
+    // - OptionT.fromOption turns an Option[A] into an OptionT[F, A]
```

Comment:
Thanks for putting in the comments - this will ensure Scala noobs like me won't get horribly confused and intimidated by the code!

Reply:
Np - always a tricky balance between commenting too little and too much